### ISSUE TITLE
Running Tools -> KeyShow causes a crash.

### ISSUE THREAD

**Issue by arkdae** [deadw00d/AROS#102]
Here is some useful debug output:

```
(gdb) bt
#0  0xb9159a23 in set_string_key (type=6 '\006', value=4160745471, qual=0 '\000')
    at /home/jlm/code/AROS_build/AROS/workbench/tools/KeyShow/./keyboardgroup_class.c:143
#1  0xb9159c35 in parse_string_key (key=0xb9190dee, type=6 '\006', value=4160745471)
    at /home/jlm/code/AROS_build/AROS/workbench/tools/KeyShow/./keyboardgroup_class.c:188
#2  0xb915a433 in read_keymap () at /home/jlm/code/AROS_build/AROS/workbench/tools/KeyShow/./keyboardgroup_class.c:343
#3  0xb915b19e in KeyboardGroup__OM_NEW (CLASS=0xb90c88b0, self=0xb90c88b0, message=0xb916b410)
    at /home/jlm/code/AROS_build/AROS/workbench/tools/KeyShow/./keyboardgroup_class.c:519
#4  0xb915e669 in KeyboardGroup_Dispatcher (__class=0xb90c88b0, __self=0xb90c88b0, __msg=0xb916b410)
    at /home/jlm/code/AROS_build/AROS/workbench/tools/KeyShow/./keyboardgroup_class.c:930
#5  0xb8dc4c0c in metaDispatcher (cl=0xb90c88b0, obj=0xb90c88b0, msg=0xb916b410)
    at /home/jlm/code/AROS_build/AROS/workbench/libs/muimaster/./support_classes.c:258
#6  0xf7b85852 in Intuition_106_NewObjectA (classPtr=0xb90c88b0, classID=0x0, tagList=0xb916b4e8, IntuitionBase=0xb88594c0)
    at /home/jlm/code/AROS_build/AROS/rom/intuition/./newobjecta.c:123
#7  0xb9158dc4 in __inline_Intuition_NewObjectA (__arg1=0xb90c88b0, __arg2=0x0, __arg3=0xb916b4e8, __IntuitionBase=0xb88594c0)
    at /home/jlm/code/AROS_build/alt-abiv0-linux-i386-d/bin/linux-i386/AROS/Development/include/inline/intuition.h:1651
#8  0xb91592d3 in main (argc=0, argv=0xb90de1c0) at /home/jlm/code/AROS_build/AROS/workbench/tools/KeyShow/./main.c:126
#9  0xb9158d0b in __startup_main (SysBase=0xb88011e0) at /home/jlm/code/AROS_build/AROS/compiler/startup/startup.c:101
#10 0xb915f917 in ___startup_entries_next (SysBase=0xb88011e0) at /home/jlm/code/AROS_build/AROS/compiler/autoinit/./programentries.c:26
#11 0xb915ed8e in __stdc_startup (SysBase=0xb88011e0) at /home/jlm/code/AROS_build/AROS/compiler/stdc/stdc_startup.c:23
#12 0xb915f917 in ___startup_entries_next (SysBase=0xb88011e0) at /home/jlm/code/AROS_build/AROS/compiler/autoinit/./programentries.c:26
#13 0xb915f5b2 in __initcommandline (SysBase=0xb88011e0) at /home/jlm/code/AROS_build/AROS/compiler/autoinit/./commandline.c:29
#14 0xb915f917 in ___startup_entries_next (SysBase=0xb88011e0) at /home/jlm/code/AROS_build/AROS/compiler/autoinit/./programentries.c:26
#15 0xb915f84b in __startup_initexit (SysBase=0xb88011e0) at /home/jlm/code/AROS_build/AROS/compiler/autoinit/./initexitsets.c:68
#16 0xb915f917 in ___startup_entries_next (SysBase=0xb88011e0) at /home/jlm/code/AROS_build/AROS/compiler/autoinit/./programentries.c:26
#17 0xb915f6ae in __startup_stdiowin (SysBase=0xb88011e0) at /home/jlm/code/AROS_build/AROS/compiler/autoinit/./stdiowin.c:28
#18 0xb915f917 in ___startup_entries_next (SysBase=0xb88011e0) at /home/jlm/code/AROS_build/AROS/compiler/autoinit/./programentries.c:26
#19 0xb915f775 in __startup_fromwb (SysBase=0xb88011e0) at /home/jlm/code/AROS_build/AROS/compiler/autoinit/./fromwb.c:21
#20 0xb915f917 in ___startup_entries_next (SysBase=0xb88011e0) at /home/jlm/code/AROS_build/AROS/compiler/autoinit/./programentries.c:26
#21 0xb9158c6e in __startup_entry (argstr=0x0, argsize=0, SysBase=0xb88011e0) at /home/jlm/code/AROS_build/AROS/compiler/startup/startup.c:82
#22 0xf7b273a1 in CallEntry (argptr=0x0, argsize=0, entry=0xb9158bb8, me=0xb9160340) at /home/jlm/code/AROS_build/AROS/rom/dos/./exit.c:129
#23 0xf7b23e35 in DosEntry () at /home/jlm/code/AROS_build/AROS/rom/dos/./createnewproc.c:752
#24 0xf7a7d517 in __inline_Exec_InitSemaphore (__arg1=0xa000, __SysBase=0x1e6de61e)
    at /home/jlm/code/AROS_build/alt-abiv0-linux-i386-d/bin/linux-i386/gen/rom/exec/exec/include/inline/exec.h:1321
#25 0xb9161850 in ?? ()
(gdb) p idx
$37 = 0 '\000'
(gdb) p str_descrs
$38 = (UBYTE *) 0xf7ffefff <error: Cannot access memory at address 0xf7ffefff>
(gdb) p value
$39 = 4160745471
(gdb) f 3
#3  0xb915b19e in KeyboardGroup__OM_NEW (CLASS=0xb90c88b0, self=0xb90c88b0, message=0xb916b410)
    at /home/jlm/code/AROS_build/AROS/workbench/tools/KeyShow/./keyboardgroup_class.c:519
519	    struct Key *key = read_keymap();
(gdb) f 2
#2  0xb915a433 in read_keymap () at /home/jlm/code/AROS_build/AROS/workbench/tools/KeyShow/./keyboardgroup_class.c:343
343	                parse_string_key(&key[i], type & KC_VANILLA, value);
(gdb) p i
$40 = 125

```
I tried to get some debug output by putting #define DEBUG 1 into both tools/mkamikeymap/writekeymap.c and workbench/libs/kms/parsekeymapseg.c, and also just removing the D() from around some lines in the latter file, but that did not cause any print output.

I think the fix is to change the for loop in read_keymap() in parsekeymapseg.c to go to 120 instead of 128. Since I could not get print outs, I am not certain, but if parsekeymapseg() is filling in the keymap, it looks like the HiKeyMap is only being filled out to 0x38 (56) instead of to 0x40 (64). The LowKeyMap is being filled to 0x40. 64+56 = 120.

When I change the loop then it no longer crashes. KeyShow seems to work correctly.

**Comment by deadw00d** [comment]
Ok, here is a catch. alt-abiv0 is based on master from 20201110 but after that date there were some fixes in master to keymaps. It's possible the the problem you are seeing is already solved in master. Can you try building master (64-bit linux hosted) and see? (I don't get a crash on my master linux hosted, but maybe I'm have some other settings than you). If the master works for you, then the fix will "flow" into alt-abiv0 at some point in time unless you need it urgently.

**Comment by arkdae** [comment]
master works just fine. The code looks very similar to what is in alt-abiv0, so my theory of the root cause must be incorrect. I certainly don't need any fix urgently. I just wanted to document things and get it stable.

I think at this point the core OS is rather stable, so I'm going to look at addon software.

**Comment by deadw00d** [comment]
FYI: It's also possible that the source file for generating the keymap had a bug and was fixed.
I'll keep this bug report open until alt-abiv0 forwards to level of master which fixes the problem

**Comment by deadw00d** [comment]
Fixed: bdae0cefbcc6b69ab821ecd3192def436db20780
